hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/impl/SignatureVerificationFutureImpl.java

```diff
@@ -0,0 +1,192 @@
+     * provided here.
+     */
+    private final Key key;
+    /**
+     * Optional: used only with hollow accounts, the EVM address associated with the hollow account. The corresponding
+     * {@link #key} was extracted from the transaction.
+     */
+    private final Account hollowAccount;
+    /**
+     * The {@link TransactionSignature} that holds the {@link Future} used to determine when the cryptographic
+     * signature check is complete,
+     */
+    private final TransactionSignature txSig;
+    /**
+     * Whether *this* future has been canceled. Used for properly implementing {@link Future} semantics.
+     */
+    private boolean canceled = false;
+
+    /**
+     * Create a new instance.
+     *
+     * @param key The key associated with this sig check. Cannot be null.
+     * @param hollowAccount The hollow account, if any
+     * @param txSig The {@link TransactionSignature}s, from which the pass/fail status of the
+     * {@link SignatureVerification} is derived. This list must contain at least one element.
+     */
+    public SignatureVerificationFutureImpl(
+            @NonNull final Key key, @Nullable final Account hollowAccount, @NonNull final TransactionSignature txSig) {
+        this.key = requireNonNull(key);
+        this.hollowAccount = hollowAccount;
+        this.txSig = requireNonNull(txSig);
+    }
+
+    /** {@inheritDoc} */
+    @Nullable
+    @Override
+    public Account hollowAccount() {
+        return hollowAccount;
+    }
+
+    /** {@inheritDoc} */
+    @NonNull
+    @Override
+    public Key key() {
+        return key;
+    }
+
+    /** {@inheritDoc} */
+    @NonNull
+    public TransactionSignature txSig() {
+        return txSig;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Attempts to cancel the {@link TransactionSignature} future. Due to the asynchronous nature of those objects,
+     * it is possible that the future has not yet even been created, and we do not currently have any way to prevent
+     * it from creating a future. The best we can do is to wait for the future to be created and then cancel it.
+     * However, waiting for the future is a blocking operation, so this would block for some time, which is not
+     * very nice.
+     *
+     * <p>instead, we can just do a "best effort" cancel, which is what we will do. If the future exists, we will
+     * cancel it. If it does not exist, we will let it be.
+     */
+    @Override
+    public boolean cancel(final boolean mayInterruptIfRunning) {
+        // If we're already done, then we can't cancel again
+        if (isDone()) {
+            return false;
+        }
+
+        // Try to cancel the underlying future, if it exists. If the underlying Future doesn't exist, then we just skip
+        // it. Best effort.
+        canceled = true;
+        final var future = txSig.getFuture();
+        if (future != null) {
+            return future.cancel(mayInterruptIfRunning);
```

Comment:
I am not sure that is right. The API on Future#cancel says:

> Attempts to cancel execution of this task. This method has no effect if the task is already completed or cancelled, or could not be cancelled for some other reason. Otherwise, if this task has not started when cancel is called, this task should never run. If the task has already started, then the mayInterruptIfRunning parameter determines whether the thread executing this task (when known by the implementation) is interrupted in an attempt to stop the task.

>The return value from this method does not necessarily indicate whether the task is now cancelled; use isCancelled.

And the definition of the return value is:

>false if the task could not be cancelled, typically because it has already completed; true otherwise. If two or more threads cause a task to be cancelled, then at least one of them returns true. Implementations may provide stronger guarantees.

If the underlying future doesn't exist yet, then as far as I'm concerned it will never exist (i.e. if it ever does exist, the caller of `SignatureVerificationFutureImpl` doesn't ever need to know that, as far as they are concerned, this future is canceled). If the underlying future does exist, then I will return whatever it reports, but that doesn't impact whether or not this future is canceled. It is canceled either way. That is how I understand the specification on Future.